pkg/volume/aws_ebs/aws_ebs.go

```diff
@@ -241,6 +242,165 @@ func (plugin *awsElasticBlockStorePlugin) ConstructVolumeSpec(volName, mountPath
+	partition := ""
+	if ebs.Partition != 0 {
+		partition = strconv.Itoa(int(ebs.Partition))
+	}
+
+	return &awsElasticBlockStoreMapper{
+		awsElasticBlockStore: &awsElasticBlockStore{
+			podUID:          podUID,
+			volName:         spec.Name(),
+			volumeID:        volumeID,
+			partition:       partition,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, spec.Name(), plugin.host)),
+		},
+		fsType:      fsType,
+		readOnly:    readOnly,
+		diskMounter: volumehelper.NewSafeFormatAndMountFromHost(plugin.GetPluginName(), plugin.host)}, nil
+}
+
+func (plugin *awsElasticBlockStorePlugin) NewBlockVolumeUnmapper(volName string, podUID types.UID) (volume.BlockVolumeUnmapper, error) {
+	return plugin.newUnmapperInternal(volName, podUID, &AWSDiskUtil{})
+}
+
+func (plugin *awsElasticBlockStorePlugin) newUnmapperInternal(volName string, podUID types.UID, manager ebsManager) (volume.BlockVolumeUnmapper, error) {
+	return &awsElasticBlockStoreUnmapper{
+		awsElasticBlockStore: &awsElasticBlockStore{
+			podUID:  podUID,
+			volName: volName,
+			manager: manager,
+			plugin:  plugin,
+		}}, nil
+}
+
+func (c *awsElasticBlockStoreUnmapper) TearDownDevice(mapPath, devicePath string) error {
+	return nil
+}
+
+type awsElasticBlockStoreUnmapper struct {
+	*awsElasticBlockStore
+}
+
+var _ volume.BlockVolumeUnmapper = &awsElasticBlockStoreUnmapper{}
+
+type awsElasticBlockStoreMapper struct {
+	*awsElasticBlockStore
+	fsType      string
+	readOnly    bool
+	diskMounter *mount.SafeFormatAndMount
+}
+
+var _ volume.BlockVolumeMapper = &awsElasticBlockStoreMapper{}
+
+func (b *awsElasticBlockStoreMapper) SetUpDevice() (string, error) {
+	return "", nil
+}
+
+// GetGlobalMapPath returns global map path and error
+// path: plugins/kubernetes.io/{PluginName}/volumeDevices/volumeID
+func (ebs *awsElasticBlockStore) GetGlobalMapPath(spec *volume.Spec) (string, error) {
+	return ebs.ebsGlobalMapPath(spec)
+}
+
+// GetPodDeviceMapPath returns pod device map path and volume name
+// path: pods/{podUid}/volumeDevices/kubernetes.io~aws
+func (ebs *awsElasticBlockStore) GetPodDeviceMapPath() (string, string) {
+	return ebs.ebsPodDeviceMapPath()
+}
+
+func (ebs *awsElasticBlockStore) ebsPodDeviceMapPath() (string, string) {
+	name := awsElasticBlockStorePluginName
+	return ebs.plugin.host.GetPodVolumeDeviceDir(ebs.podUID, kstrings.EscapeQualifiedNameForDisk(name)), ebs.volName
+}
+
+func (ebs *awsElasticBlockStore) ebsGlobalMapPath(spec *volume.Spec) (string, error) {
+	return makeVDPDNameInternal(ebs.plugin.host, string(ebs.volumeID)), nil
+}
+
+// make a directory like /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/volumeDevices/volumeID
+func makeVDPDNameInternal(host volume.VolumeHost, vID string) string {
+	return path.Join(host.GetVolumeDevicePluginDir(awsElasticBlockStorePluginName), vID)
+}
+
+func getVolumeIDFromGlobalVDPVPath(globalPath string) (string, error) {
```

Comment:
Reading the code, it seems that `filepath.Base()` does something very similar.

Reply:
ack